skimage/filters/ridges.py

```diff
@@ -407,66 +313,40 @@ def frangi(image, sigmas=range(1, 10, 2), scale_range=None,
-        # structure see equations (11) and (15) in reference [1]_
-        r_a = np.inf if ndim == 2 else _divide_nonzero(*lambdas) ** 2
-
-        # Compute sensitivity to deviation from a blob-like structure,
-        # see equations (10) and (15) in reference [1]_,
-        # np.abs(lambda2) in 2D, np.sqrt(np.abs(lambda2 * lambda3)) in 3D
-        filtered_raw = np.abs(np.multiply.reduce(lambdas)) ** (1/len(lambdas))
-        r_b = _divide_nonzero(lambda1, filtered_raw) ** 2
-
-        # Compute sensitivity to areas of high variance/texture/structure,
-        # see equation (12)in reference [1]_
-        r_g = sum([lambda1 ** 2] + [lambdai ** 2 for lambdai in lambdas])
-
-        # Compute output image for given (sigma) scale and store results in
-        # (n+1)D matrices, see equations (13) and (15) in reference [1]_
-        filtered_array[i] = ((1 - np.exp(-r_a / alpha_sq))
-                             * np.exp(-r_b / beta_sq)
-                             * (1 - np.exp(-r_g / gamma_sq)))
-
-        lambdas_array[i] = np.max(lambdas, axis=0)
-
-    # Remove background
-    filtered_array[lambdas_array > 0] = 0
-
-    # Return for every pixel the maximum value over all (sigma) scales
-    return np.max(filtered_array, axis=0)
+    check_nD(image, [2, 3])  # Check image dimensions.
+    image = image.astype(_supported_float_type(image.dtype), copy=False)
+    if not black_ridges:  # Normalize to black ridges.
+        image = -image
+
+    filtered = []
+    for i, sigma in enumerate(sigmas):  # Filter for all sigmas.
+        eigvals = hessian_matrix_eigvals(hessian_matrix(
+            image, sigma, mode=mode, cval=cval, use_gaussian_derivatives=True))
+        # Sort eigenvalues by magnitude.
+        eigvals = np.take_along_axis(eigvals, abs(eigvals).argsort(0), 0)
+        lambda1 = eigvals[0]
+        if image.ndim == 2:
+            lambda2, = np.maximum(eigvals[1:], 1e-10)
+            r_a = np.inf  # implied by eq. (15).
+            r_b = abs(lambda1) / lambda2  # eq. (15).
+        else:  # ndim == 3
+            lambda2, lambda3 = np.maximum(eigvals[1:], 1e-10)
+            r_a = lambda2 / lambda3  # eq. (11).
+            r_b = abs(lambda1) / np.sqrt(lambda2 * lambda3)  # eq. (10).
+        s = np.sqrt((eigvals ** 2).sum(0))  # eq. (12).
+        if gamma is None:
+            gamma = s.max() / 2
+            if gamma == 0:
+                gamma = 1  # If s == 0 everywhere, gamma doesn't matter.
+        # Filtered image, eq. (13) and (15).  Our implementation relies on the
+        # blobness exponential factor underflowing to zero whenever the second
+        # or third eigenvalues are negative (we clip them to 1e-10, to make r_b
+        # very large).
+        vals = ((1 - np.exp(-r_a**2 / (2 * alpha**2)))  # plate sensitivity.
+                * np.exp(-r_b**2 / (2 * beta**2))  # blobness.
+                * (1 - np.exp(-s**2 / (2 * gamma**2))))  # structuredness.
```

Comment:
Let's split this into a few separate inplace calls. That should retain readability, but also may help reduce memory overhead. 

```suggestion
        vals = 1 - np.exp(-r_a**2 / (2 * alpha**2)  # plate sensitivity
        vals *= np.exp(-r_b**2 / (2 * beta**2))  # blobness
        vals *= (1 - np.exp(-s**2 / (2 * gamma**2)))  # structuredness
```

Reply:
sure.